New instrument: HP8657B RF signal generator

LongnoseRob opened the pull request:
Simple RF signal generator implementation, legacy device with listenm-only GPIB/HPIB interface

bmoneke commented on pymeasure/instruments/hp/hp8657b.py:
```diff
@@ -0,0 +1,190 @@
+    with the instrument.
+    """
+
+    def __init__(self, adapter, **kwargs):
+        kwargs.setdefault('read_termination', '\r\n')
+        kwargs.setdefault('send_end', True)
+        super().__init__(
+            adapter,
+            "Hewlett-Packard HP8657B",
+            includeSCPI=False,
+            **kwargs,
+        )
```
If the device does not support a different termination/send_end, I'd put these kwargs direct in the super.init:
```suggestion
    def __init__(self, adapter, **kwargs):
        super().__init__(
            adapter,
            "Hewlett-Packard HP8657B",
            includeSCPI=False,
            read_termination='\r\n',
            send_end=True,
            **kwargs,
        )
```

bmoneke commented on pymeasure/instruments/hp/hp8657b.py:
```diff
@@ -0,0 +1,190 @@
+            **kwargs,
+        )
+
+    def check_errors(self):
+        """
+        Method to read the error status register
+
+        as the 8657B does not support any readout of values, this will return 0 and log a warning
+
+        """
+        log.warning("HP8657B Does not support error status readout")
+        return 0
```
Why do you return 0? The user will think, there is no error. Wouldn't it be better to raise an Exception?

bmoneke commented on pymeasure/instruments/hp/hp8657b.py:
```diff
@@ -0,0 +1,190 @@
+    id = "HP 8657B Signal generator"  #: Manual ID entry, as the instrument does not talk with us
+
+    # List of modulation sources
+    modulation_source = {"External": 1,
+                         "Int_400Hz": 2,
+                         "Int_1000Hz": 3,
+                         "OFF": 4,
+                         "DC_FM": 5}
```
If you use an enum, the user will have autocompletion in their editor, e.g. `inst.am_source = Modulation.OFF`

LongnoseRob replied:
I like the idea, 
however 'Instrument.control()' does not accecpt Enums as inputs.
Or did I miss a change regarding this?

bmoneke replied:
If you use an IntEnum, the enumeration member will be interpreted as an integer, whenever an integer is expected.
I did not test it yet, but it should work, as `"Command %i" % Modulation.OFF` will be `"Command 4"`, if 
```
class Modulation(IntEnum):
    OFF = 4
```

LongnoseRob replied:
Thanks, it was the `map_values=True` which does not like the IntEnum.
But as this unit is listen-only we do not `map_values`

bmoneke commented:
Thanks for your pull request.

Please provide at least a few [protocol tests](https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments.html#protocol-tests) for this device.

bmoneke commented on pymeasure/instruments/hp/hp8657b.py:
```diff
@@ -0,0 +1,188 @@
+    am_depth = Instrument.setting(
+        "AM %2.1f PC",
+        """
+        sets the modulation depth for AM,
+        usable range 0-99.9%
```
The property docstrings should follow the normal docstring style (according to the not yet merged documentation overhaul). Therefore the first sentence is an imperative (and starts with a capital letter):
```suggestion
        Set the modulation depth for AM,
        usable range 0-99.9%
```

Additionally, I'd add the abbreviation definitions of "AM" and "FM" to the device docstring, just in case. I guess they mean amplitude and frequency modulation.

LongnoseRob replied:
added a shot glossary with https://github.com/pymeasure/pymeasure/pull/732/commits/6d61a9ae93fb0335ed54913de60c07202716f59d

msmttchr commented on pymeasure/instruments/hp/hp8657b.py:
```diff
@@ -0,0 +1,200 @@
+        """
+        self.adapter.connection.clear()
+
+    id = "HP 8657B Signal generator"  #: Manual ID entry, as the instrument does not talk with us
```
@LongnoseRob, nice idea to have an `id` anyway for uniformity with other scpi instruments.
I would consider to add the string in IEEE 488.2 format.
I report here an excerpt from a documentation for an Anritsu instrument.

`identification query. query only. this query returns an instrument identification string in ieee-488.2 specified <arbitrary ascii> format consisting of four fields separated by commas. the fields are: <manufacturer>, <model>, <serial #>,<firmware revision level>`

So, in your case, it should be something like:
`"HP,8657B,N/A,N/A"`

What do you think about this?
If there is a general agreement, we could put this as suggestion in PR #712.

LongnoseRob replied:
> So, in your case, it should be something like:
"HP,8657B,N/A,N/A"
 What do you think about this?

Thanks for the feedback and the improvement idea.
One question the _implementation_, would t 
```
id = "HP,8657B,N/A,N/A"
```
be sufficient or should it be 

```
def id(self):
    """"FAKE *IDN? return"""
    return "HP,8657B,N/A,N/A"
```

msmttchr replied:
Well, I think that property exists to allow access to a value that requires some calculation (i.e. code execution) like an attribute. For example:
```
@property
    def id(self):
        """ Requests and returns the identification of the instrument. """
        if self.SCPI:
            return self.ask("*IDN?").strip()
        else:
            raise NotImplementedError("Non SCPI instruments require implementation in subclasses")
```
In our case, we don't need to run any code, so we can go back to plain attribute assignment.

bmoneke commented on docs/api/instruments/hp/hp8657B.rst:
```diff
@@ -4,9 +4,20 @@ HP Signal generator HP8657B
+    * This instrument does not support reading back values, as it is a listen-only GPIB device
     
-    * Other instrument of this family could be implemented using the dymaic ranges feature
+    * Other instruments of this family could be implemented using the dymaic ranges feature
+    
+    * Optorional pulse modulation feature is not support yet
```
```suggestion
    * Other instruments of this family could be implemented using the dynamic ranges feature.
    
    * Optorional pulse modulation feature is not supported yet.
```
What does "Optorional" mean?

bmoneke commented on pymeasure/instruments/hp/hp8657b.py:
```diff
@@ -162,9 +172,9 @@ def clear(self):
+        "AO %g DB",
         """
-        sets the output offset in dB, usable range -199 to +199 dB
+        Sets the output offset in dB, usable range -199 to +199 dB
```
```suggestion
        Set the output offset in dB, usable range -199 to +199 dB.
```

bmoneke commented on pymeasure/instruments/hp/hp8657b.py:
```diff
@@ -174,7 +184,7 @@ def clear(self):
         "R%d",
         """
-        controls if the output is enabled
+        Controls the output
```
```suggestion
        Control whether the output is enabled.
```

bmoneke commented:
A few more comments, mainly regarding docstring formulation.

bmoneke commented on pymeasure/instruments/hp/hp8657b.py:
```diff
@@ -0,0 +1,200 @@
+        values=[0, 99.9],
+        )
+
+    am_source = Instrument.setting(
+        "AM S%i",
+        """
+        Set the source for the AM function
+
+        ==========  =======
+        Value       Meaning
+        ==========  =======
+        OFF         no modulation active
+        Int_400Hz   internal 400 Hz modulation source
+        Int_1000Hz  internal 1000 Hz modulation source
+        External    External source, AC coupling
+        ==========  =======
+
+        *Note:*
+            * AM & FM can be active at the same time
+            * only one internal source can be active at the time
+            * use "OFF" to deactivate AM
+
+        """,
+        validator=strict_discrete_set,
+        values=Modulation,
+        )
```
Please give the user a hint, that he might use the `Modulation` enum (similarly below). Otherwise the user will never know.

LongnoseRob replied:
I will add a codeblock exmaplew how to use modulations

bmoneke reviewed (changes requested):
Having implemented that small doc change, it will look good to me.
@msmttchr do you have anything else?

msmttchr commented:
> Having implemented that small doc change, it will look good to me. @msmttchr do you have anything else?

I agree. It looks good to me as well. Thanks

msmttchr reviewed (approved):
@LongnoseRob thanks for your contribution. It looks good to me.
@bmoneke feel free to merge after it is good for you.

bmoneke commented on pymeasure/instruments/hp/hp8657b.py:
```diff
@@ -0,0 +1,222 @@
+    am_source = Instrument.setting(
+        "AM S%i",
+        """
+        Set the source for the AM function
+
+        ==========  =======
+        Value       Meaning
+        ==========  =======
+        OFF         no modulation active
+        Int_400Hz   internal 400 Hz modulation source
+        Int_1000Hz  internal 1000 Hz modulation source
+        External    External source, AC coupling
+        ==========  =======
+
+        *Note:*
+            * AM & FM can be active at the same time
+            * only one internal source can be active at the time
+            * use "OFF" to deactivate AM
+
+        usage example:
+
+        .. code-block:: python
+
+            sig_gen = HP8657B("GPIB::7")
+            ...
+            sig_gen.am_source = sig_gen.Modulation.Int_400Hz    #  Enable int. 400 Hz source for AM
+            sig_gen.am_depth = 50                               #  Set AM modulation depth to 50%
+            ...
+            sig_gen.am_source = sig_gen.Modulation.OFF          #  Turn AM off
```
```suggestion
        Set the source for the AM function with :attr:`Modulation` enumeration.

        ==========  =======
        Value       Meaning
        ==========  =======
        OFF         no modulation active
        INT_400HZ   internal 400 Hz modulation source
        INT_1000HZ  internal 1000 Hz modulation source
        EXTERNAL    External source, AC coupling
        ==========  =======

        *Note:*
            * AM & FM can be active at the same time
            * only one internal source can be active at the time
            * use "OFF" to deactivate AM

        usage example:

        .. code-block:: python

            sig_gen = HP8657B("GPIB::7")
            ...
            sig_gen.am_source = sig_gen.Modulation.INT_400HZ    #  Enable int. 400 Hz source for AM
            sig_gen.am_depth = 50                               #  Set AM modulation depth to 50%
            ...
            sig_gen.am_source = sig_gen.Modulation.OFF          #  Turn AM off
```

bmoneke reviewed (commented):
I suggest some changes (made more clear how to set the values and adjusted the docstrings to the Enum).

Thank you for developing the usage of enums together with me. As it is a new usage, treading new paths requires some work, thanks for staying with it.
With your hints, we added it into the documentation and will improve their usage. Thanks for the feedback!

LongnoseRob commented:
> I suggest some changes (made more clear how to set the values and adjusted the docstrings to the Enum).
> 
> Thank you for developing the usage of enums together with me. As it is a new usage, treading new paths requires some work, thanks for staying with it. With your hints, we added it into the documentation and will improve their usage. Thanks for the feedback!

Thank you for the code-review and the constructive critique!.

bilderbuchi commented on pymeasure/instruments/hp/hp8657b.py:
```diff
@@ -0,0 +1,222 @@
+    def shutdown(self):
+        self.output_enabled = False
+        self.adapter.connection.clear()
+        self.adapter.connection.close()
```
Arguably, this should/could call `super().shutdown()`, too (although, cf #513). @bmoneke

bmoneke replied:
Oh, yes...

LongnoseRob replied:
Should I add this line `super().shutdown()`  and push again?
O should we  check all non-SCPI implementations and check if `shutdown()` has this line?

bmoneke replied:
The pull request is already merged. I can just add this line directly.

LongnoseRob replied:
Ok, thats also fine with me.
Thanks

bilderbuchi replied:
The PR has been merged, so you'd need to open another PR for that.
Arguably, all instruments that have redefined `shutdown` should call the parent method, too (or have a good/documented-by-comment reason not to). _That_ would be a useful PR, if you want to. :-)

LongnoseRob replied:
> The PR has been merged, so you'd need to open another PR for that. Arguably, all instruments that have redefined `shutdown` should call the parent method, too (or have a good/documented-by-comment reason not to). _That_ would be a useful PR, if you want to. :-)

A quick greo only gave about 20 instruments with `includeSCPI=False`, so Ill open a new PR for that
